File: packages/sdk-python/cred/identity.py
"""Agent Identity — DID:key generation and management.

Uses Ed25519 key pairs with did:key encoding (RFC draft-multiformats-multibase).
Uses `cryptography` package for Ed25519 operations.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives import serialization

# ── Cred Public Key (pinned at build time) ───────────────────────────────────

CRED_PUBLIC_KEY_HEX = "PLACEHOLDER_REPLACE_BEFORE_LAUNCH"

# ── Base58btc alphabet (Bitcoin) ─────────────────────────────────────────────

_BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def _encode_base58(data: bytes) -> str:
    """Encode bytes to base58btc string."""
    if len(data) == 0:
        return ""

    # Count leading zeros
    zeros = 0
    for byte in data:
        if byte != 0:
            break
        zeros += 1

    # Convert to big integer and repeatedly divide by 58
    size = (len(data) * 138 // 100) + 1
    b58 = [0] * size
    length = 0

    for byte in data:
        carry = byte
        i = 0
        for j in range(size - 1, -1, -1):
            if carry == 0 and i >= length:
                break
            carry += 256 * b58[j]
            b58[j] = carry % 58
            carry //= 58
            i += 1
        length = i

    # Skip leading zeros in base58 result
    start = size - length
    while start < size and b58[start] == 0:
        start += 1

    # Build result string
    result = "1" * zeros
    for i in range(start, size):
        result += _BASE58_ALPHABET[b58[i]]

    return result


def _decode_base58(s: str) -> bytes:
    """Decode base58btc string to bytes."""
    if len(s) == 0:
        return b""

    # Build alphabet lookup
    lookup = {char: i for i, char in enumerate(_BASE58_ALPHABET)}

    # Count leading '1's (zeros)
    zeros = 0
    for char in s:
        if char != "1":
            break
        zeros += 1

    # Decode to bytes
    size = (len(s) * 733 // 1000) + 1
    result = [0] * size
    length = 0

    for char in s:
        value = lookup.get(char)
        if value is None:
            raise ValueError(f"Invalid base58 character: {char}")

        carry = value
        i = 0
        for j in range(size - 1, -1, -1):
            if carry == 0 and i >= length:
                break
            carry += 58 * result[j]
            result[j] = carry % 256
            carry //= 256
            i += 1
        length = i

    # Skip leading zeros in byte array
    start = size - length
    while start < size and result[start] == 0:
        start += 1

    # Prepend zeros and return
    return bytes([0] * zeros + result[start:])


# ── Ed25519 multicodec prefix ────────────────────────────────────────────────

# did:key multicodec prefix for Ed25519 public keys: 0xed01
# Encoded as varint: [0xed, 0x01]
_ED25519_MULTICODEC_PREFIX = bytes([0xED, 0x01])


# ── Types ────────────────────────────────────────────────────────────────────


@dataclass
class ExportedIdentity:
    """Exported identity for persistence."""

    did: str
    private_key_hex: str


class AgentIdentity:
    """Agent identity with Ed25519 key pair and DID.

    Attributes:
        did: DID in did:key format: did:key:z6Mk<base58-encoded-public-key>
        public_key: Raw 32-byte Ed25519 public key
        private_key: Raw 32-byte Ed25519 private key
    """

    def __init__(
        self,
        did: str,
        public_key: bytes,
        private_key: bytes,
    ) -> None:
        self._did = did
        self._public_key = bytes(public_key)
        self._private_key = bytes(private_key)

    @property
    def did(self) -> str:
        return self._did

    @property
    def public_key(self) -> bytes:
        return bytes(self._public_key)

    @property
    def private_key(self) -> bytes:
        return bytes(self._private_key)

    def export(self) -> dict[str, str]:
        """Export identity for persistence.

        Returns:
            Dict with 'did' and 'private_key_hex' keys.
        """
        return {
            "did": self._did,
            "private_key_hex": self._private_key.hex(),
        }

    @classmethod
    def from_export(cls, data: dict[str, str]) -> "AgentIdentity":
        """Import a persisted agent identity.

        Args:
            data: Dict with 'did' and 'private_key_hex' keys.

        Returns:
            AgentIdentity with full key pair.

        Raises:
            ValueError: If DID format is invalid or private key is wrong size.
        """
        did = data["did"]
        private_key_hex = data["private_key_hex"]

        # Validate DID format
        if not did.startswith("did:key:z"):
            raise ValueError("Invalid DID format: must start with did:key:z")

        # Decode private key from hex
        private_key_bytes = bytes.fromhex(private_key_hex)
        if len(private_key_bytes) != 32:
            raise ValueError("Invalid private key: must be 32 bytes")

        # Derive public key from private key using cryptography
        private_key_obj = Ed25519PrivateKey.from_private_bytes(private_key_bytes)
        public_key_obj = private_key_obj.public_key()
        public_key_bytes = public_key_obj.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

        # Verify DID matches derived public key
        derived_did = _public_key_to_did(public_key_bytes)
        if derived_did != did:
            raise ValueError("DID does not match derived public key")

        return cls(did, public_key_bytes, private_key_bytes)


def _public_key_to_did(public_key: bytes) -> str:
    """Convert raw Ed25519 public key to did:key format."""
    # Prepend multicodec prefix and encode as base58btc
    prefixed = _ED25519_MULTICODEC_PREFIX + public_key
    # 'z' is the multibase prefix for base58btc
    return f"did:key:z{_encode_base58(prefixed)}"


def generate_agent_identity() -> AgentIdentity:
    """Generate a new agent identity with Ed25519 key pair.

    Returns:
        AgentIdentity with DID, public key, private key, and export function.
    """
    # Generate new Ed25519 key pair
    private_key_obj = Ed25519PrivateKey.generate()
    public_key_obj = private_key_obj.public_key()

    # Extract raw key bytes
    private_key_bytes = private_key_obj.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_key_bytes = public_key_obj.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )

    # Build did:key from public key
    did = _public_key_to_did(public_key_bytes)

    return AgentIdentity(did, public_key_bytes, private_key_bytes)


# ── Receipt Verification ─────────────────────────────────────────────────────


def _base64url_decode(s: str) -> bytes:
    """Decode base64url string to bytes."""
    import base64

    # Replace URL-safe characters and add padding
    s = s.replace("-", "+").replace("_", "/")
    # Add padding
    padding = 4 - (len(s) % 4)
    if padding != 4:
        s += "=" * padding
    return base64.b64decode(s)


def verify_delegation_receipt(
    receipt: Optional[str],
    expected_did: str,
    cred_public_key: Optional[str] = None,
) -> bool:
    """Verify a delegation receipt from the Cred API.

    Receipts are JWS compact serialization (header.payload.signature) signed
    by Cred's Ed25519 key. Returns True if signature is valid and DID matches.

    Args:
        receipt: JWS compact serialization string (or None).
        expected_did: Expected agent DID (sub claim in receipt payload).
        cred_public_key: Cred's public key in hex format (defaults to CRED_PUBLIC_KEY_HEX).

    Returns:
        True if receipt is valid and matches expected DID, False otherwise.

    Raises:
        ValueError: If CRED_PUBLIC_KEY_HEX is still placeholder.
    """
    # Return False for missing receipts (don't throw)
    if not receipt:
        return False

    cred_public_key_hex = cred_public_key or CRED_PUBLIC_KEY_HEX

    # Placeholder check — can't verify without real key
    if cred_public_key_hex == "PLACEHOLDER_REPLACE_BEFORE_LAUNCH":
        raise ValueError("CRED_PUBLIC_KEY_HEX is placeholder — cannot verify receipts")

    try:
        # Parse JWS compact serialization: header.payload.signature
        parts = receipt.split(".")
        if len(parts) != 3:
            return False

        header_b64, payload_b64, signature_b64 = parts

        # Decode and validate header
        header = json.loads(_base64url_decode(header_b64).decode("utf-8"))
        if header.get("alg") != "EdDSA" or header.get("typ") != "JWT":
            return False

        # Decode payload and verify DID matches
        payload = json.loads(_base64url_decode(payload_b64).decode("utf-8"))
        if payload.get("sub") != expected_did:
            return False

        # Verify signature
        signature_input = f"{header_b64}.{payload_b64}".encode("utf-8")
        signature = _base64url_decode(signature_b64)

        # Build Ed25519 public key object from raw bytes
        cred_public_key_bytes = bytes.fromhex(cred_public_key_hex)
        if len(cred_public_key_bytes) != 32:
            return False

        public_key_obj = Ed25519PublicKey.from_public_bytes(cred_public_key_bytes)
        public_key_obj.verify(signature, signature_input)
        return True
    except Exception:
        return False

File: packages/sdk/src/identity.ts
/**
 * Agent Identity — DID:key generation and management
 *
 * Uses Ed25519 key pairs with did:key encoding (RFC draft-multiformats-multibase).
 * Zero external dependencies — Node.js crypto module only.
 */

import { generateKeyPair, KeyObject, verify, createPrivateKey, createPublicKey } from 'node:crypto';
import { promisify } from 'node:util';

const generateKeyPairAsync = promisify(generateKeyPair);

// ── Cred Public Key (pinned at build time) ───────────────────────────────────

/**
 * Cred's Ed25519 public key for verifying delegation receipts.
 */
export const CRED_PUBLIC_KEY_HEX = 'ed50f12caeed3e3054be071c2884fda47752a02b5d98e37dfd79db5fd502699f';

// Default max age for legacy receipts (no `exp` claim), matching the
// server's default receipt TTL. Allowed clock skew when checking `exp`.
const DEFAULT_RECEIPT_MAX_AGE_SECONDS = 3600;
const RECEIPT_CLOCK_SKEW_SECONDS = 60;

// ── Types ────────────────────────────────────────────────────────────────────

export type AgentStatus = 'active' | 'suspended' | 'revoked';

export interface GenerateIdentityOptions {
  scopeCeiling?: string[];
  status?: AgentStatus;
}

export interface AgentIdentity {
  /** DID in did:key format: did:key:z6Mk<base58-encoded-public-key> */
  did: string;
  /** Raw 32-byte Ed25519 public key */
  publicKey: Uint8Array;
  /** Raw 32-byte Ed25519 private key (caller must persist securely) */
  privateKey: Uint8Array;
  /** Maximum scopes this agent can ever request (empty = uncapped) */
  scopeCeiling: string[];
  /** Agent status — revoked agents cannot receive delegations */
  status: AgentStatus;
  /** Export identity for persistence */
  export(): ExportedIdentity;
}

export interface ExportedIdentity {
  did: string;
  privateKeyHex: string;
  scopeCeiling: string[];
  status: AgentStatus;
}

export interface ImportParams {
  did: string;
  privateKeyHex: string;
  scopeCeiling?: string[];
  status?: AgentStatus;
}

export interface VerifyReceiptOptions {
  /** Expected agent DID (sub claim in receipt payload) */
  expectedDid: string;
  /** Cred's public key in hex format (defaults to CRED_PUBLIC_KEY_HEX) */
  credPublicKey?: string;
  /**
   * Max age (seconds) allowed for legacy receipts that predate the `exp`
   * claim (falls back to `iat` + this value). Defaults to 3600 (1 hour) to
   * match the server's default receipt TTL. Ignored for receipts that carry
   * an `exp` claim — those are checked against `exp` directly.
   */
  maxAgeSeconds?: number;
}

export interface DelegationReceiptPayload {
  /** Issuer — Cred's DID (did:web:cred.ninja) */
  iss: string;
  /** Subject — the requesting agent's DID */
  sub: string;
  /** Issued at (Unix timestamp) */
  iat: number;
  /** Service slug */
  service: string;
  /** Granted scopes */
  scopes: string[];
  /** Pseudonymous user ID (hashed) */
  userId: string;
  /** App client ID */
  appClientId: string;
  /** Delegation identifier */
  delegationId?: string;
  /** Parent delegation identifier for sub-delegation chains */
  parentDelegationId?: string;
  /** Root delegation starts at 0 and increments per hop */
  chainDepth?: number;
  /** SHA-256 of the parent receipt when this is a child delegation */
  parentReceiptHash?: string;
  /** Expiry (Unix timestamp). Absent on legacy receipts minted before this claim existed. */
  exp?: number;
  /** Intended audience for the receipt (e.g. the issuing server's base URI) */
  aud?: string;
}

// ── Base58btc alphabet (Bitcoin) ─────────────────────────────────────────────

const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

function encodeBase58(bytes: Uint8Array): string {
  if (bytes.length === 0) return '';

  // Count leading zeros
  let zeros = 0;
  for (const byte of bytes) {
    if (byte !== 0) break;
    zeros++;
  }

  // Convert to big integer and repeatedly divide by 58
  const size = Math.ceil(bytes.length * 138 / 100) + 1;
  const b58 = new Uint8Array(size);
  let length = 0;

  for (const byte of bytes) {
    let carry = byte;
    let i = 0;
    for (let j = size - 1; (carry !== 0 || i < length) && j >= 0; j--, i++) {
      carry += 256 * b58[j];
      b58[j] = carry % 58;
      carry = Math.floor(carry / 58);
    }
    length = i;
  }

  // Skip leading zeros in base58 result
  let start = size - length;
  while (start < size && b58[start] === 0) {
    start++;
  }

  // Build result string
  let result = '1'.repeat(zeros);
  for (let i = start; i < size; i++) {
    result += BASE58_ALPHABET[b58[i]];
  }

  return result;
}

// ── Ed25519 multicodec prefix ────────────────────────────────────────────────

// did:key multicodec prefix for Ed25519 public keys: 0xed01
// Encoded as varint: [0xed, 0x01]
const ED25519_MULTICODEC_PREFIX = new Uint8Array([0xed, 0x01]);

// ── Core Functions ───────────────────────────────────────────────────────────

/**
 * Generate a new agent identity with Ed25519 key pair.
 *
 * @returns AgentIdentity with DID, public key, private key, and export function
 */
export async function generateAgentIdentity(
  opts?: GenerateIdentityOptions,
): Promise<AgentIdentity> {
  const { publicKey, privateKey } = await generateKeyPairAsync('ed25519');

  // Extract raw key bytes from KeyObjects
  const publicKeyBytes = extractRawPublicKey(publicKey);
  const privateKeyBytes = extractRawPrivateKey(privateKey);

  // Build did:key from public key
  const did = publicKeyToDid(publicKeyBytes);

  const scopeCeiling = opts?.scopeCeiling ?? [];
  const status: AgentStatus = opts?.status ?? 'active';

  return createIdentity(did, publicKeyBytes, privateKeyBytes, scopeCeiling, status);
}

/**
 * Import a persisted agent identity.
 *
 * @param params - The exported identity (DID + private key hex)
 * @returns AgentIdentity with full key pair
 */
export function importAgentIdentity(params: ImportParams): AgentIdentity {
  const { did, privateKeyHex, scopeCeiling, status } = params;

  // Validate DID format
  if (!did.startsWith('did:key:z')) {
    throw new Error('Invalid DID format: must start with did:key:z');
  }

  // Decode private key from hex
  const privateKeyBytes = hexToBytes(privateKeyHex);
  if (privateKeyBytes.length !== 32) {
    throw new Error('Invalid private key: must be 32 bytes');
  }

  // Derive public key from private key using Node.js crypto
  const privateKeyObject = createPrivateKey({
    key: Buffer.concat([
      // Ed25519 PKCS#8 prefix
      Buffer.from('302e020100300506032b657004220420', 'hex'),
      Buffer.from(privateKeyBytes),
    ]),
    format: 'der',
    type: 'pkcs8',
  });
  const publicKeyObject = createPublicKey(privateKeyObject);
  const publicKeyBytes = extractRawPublicKey(publicKeyObject);

  // Verify DID matches derived public key
  const derivedDid = publicKeyToDid(publicKeyBytes);
  if (derivedDid !== did) {
    throw new Error('DID does not match derived public key');
  }

  return createIdentity(did, publicKeyBytes, privateKeyBytes, scopeCeiling ?? [], status ?? 'active');
}

/**
 * Verify a delegation receipt from the Cred API.
 *
 * Receipts are JWS compact serialization (header.payload.signature) signed
 * by Cred's Ed25519 key. Returns true if signature is valid and DID matches.
 *
 * @param receipt - JWS compact serialization string (or null/undefined)
 * @param opts - Verification options
 * @returns true if receipt is valid and matches expected DID
 */
export async function verifyDelegationReceipt(
  receipt: string | null | undefined,
  opts: VerifyReceiptOptions,
): Promise<boolean> {
  // Return false for missing receipts (don't throw)
  if (!receipt) {
    return false;
  }

  const credPublicKeyHex = opts.credPublicKey ?? CRED_PUBLIC_KEY_HEX;

  // Placeholder check — can't verify without real key
  if (credPublicKeyHex === 'PLACEHOLDER_REPLACE_BEFORE_LAUNCH') {
    throw new Error('CRED_PUBLIC_KEY_HEX is placeholder — cannot verify receipts');
  }

  try {
    // Parse JWS compact serialization: header.payload.signature
    const parts = receipt.split('.');
    if (parts.length !== 3) {
      return false;
    }

    const [headerB64, payloadB64, signatureB64] = parts;

    // Decode and validate header
    const header = JSON.parse(base64UrlDecode(headerB64).toString('utf8'));
    if (header.alg !== 'EdDSA' || header.typ !== 'JWT') {
      return false;
    }

    // Verify signature FIRST, before trusting anything in the payload —
    // defense-in-depth: an attacker-controlled/unsigned payload should
    // never influence a claims/expiry decision. Only decode and act on the
    // payload once the signature is confirmed to have come from
    // `credPublicKeyHex`.
    const signatureInput = Buffer.from(`${headerB64}.${payloadB64}`, 'utf8');
    const signature = base64UrlDecode(signatureB64);

    // Build Ed25519 public key object from raw bytes
    const credPublicKeyBytes = hexToBytes(credPublicKeyHex);
    if (credPublicKeyBytes.length !== 32) {
      return false;
    }

    const publicKeyObject = createPublicKey({
      key: Buffer.concat([
        // Ed25519 SPKI prefix
        Buffer.from('302a300506032b6570032100', 'hex'),
        Buffer.from(credPublicKeyBytes),
      ]),
      format: 'der',
      type: 'spki',
    });

    const signatureValid = verify(null, signatureInput, publicKeyObject, signature);
    if (!signatureValid) {
      return false;
    }

    // Only now decode and trust claims from the payload.
    const payload: DelegationReceiptPayload = JSON.parse(
      base64UrlDecode(payloadB64).toString('utf8'),
    );
    if (payload.sub !== opts.expectedDid) {
      return false;
    }

    // Expiry check. Receipts with an `exp` claim are rejected once they age
    // past it (with a small allowance for clock skew). Legacy receipts
    // (minted before `exp` existed) fall back to `iat` + maxAgeSeconds, so
    // they age out rather than remaining valid forever.
    const nowSeconds = Math.floor(Date.now() / 1000);
    const maxAgeSeconds = opts.maxAgeSeconds ?? DEFAULT_RECEIPT_MAX_AGE_SECONDS;
    if (typeof payload.exp === 'number') {
      if (nowSeconds > payload.exp + RECEIPT_CLOCK_SKEW_SECONDS) {
        return false;
      }
    } else if (typeof payload.iat === 'number') {
      if (nowSeconds > payload.iat + maxAgeSeconds) {
        return false;
      }
    }

    return true;
  } catch {
    return false;
  }
}

// ── Internal Helpers ─────────────────────────────────────────────────────────

function createIdentity(
  did: string,
  publicKey: Uint8Array,
  privateKey: Uint8Array,
  scopeCeiling: string[],
  status: AgentStatus,
): AgentIdentity {
  // Store copies internally
  const pubKeyCopy = new Uint8Array(publicKey);
  const privKeyCopy = new Uint8Array(privateKey);
  const ceilCopy = [...scopeCeiling];

  return {
    did,
    scopeCeiling: ceilCopy,
    status,
    // Return fresh copies to prevent accidental mutation
    get publicKey(): Uint8Array {
      return new Uint8Array(pubKeyCopy);
    },
    get privateKey(): Uint8Array {
      return new Uint8Array(privKeyCopy);
    },
    export(): ExportedIdentity {
      return {
        did,
        privateKeyHex: bytesToHex(privKeyCopy),
        scopeCeiling: [...ceilCopy],
        status,
      };
    },
  };
}

function publicKeyToDid(publicKey: Uint8Array): string {
  // Prepend multicodec prefix and encode as base58btc
  const prefixed = new Uint8Array(ED25519_MULTICODEC_PREFIX.length + publicKey.length);
  prefixed.set(ED25519_MULTICODEC_PREFIX);
  prefixed.set(publicKey, ED25519_MULTICODEC_PREFIX.length);

  // 'z' is the multibase prefix for base58btc
  return `did:key:z${encodeBase58(prefixed)}`;
}

function extractRawPublicKey(keyObject: KeyObject): Uint8Array {
  // Export as SubjectPublicKeyInfo (SPKI) DER format
  const spki = keyObject.export({ type: 'spki', format: 'der' });
  // Ed25519 SPKI is 44 bytes: 12-byte header + 32-byte key
  // The raw key is the last 32 bytes
  return new Uint8Array(spki.slice(-32));
}

function extractRawPrivateKey(keyObject: KeyObject): Uint8Array {
  // Export as PKCS#8 DER format
  const pkcs8 = keyObject.export({ type: 'pkcs8', format: 'der' });
  // Ed25519 PKCS#8 is 48 bytes: 16-byte header + 32-byte key
  // The raw key is the last 32 bytes
  return new Uint8Array(pkcs8.slice(-32));
}

function bytesToHex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('hex');
}

function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new Error('Invalid hex string: odd length');
  }
  return new Uint8Array(Buffer.from(hex, 'hex'));
}

function base64UrlDecode(str: string): Buffer {
  // Replace URL-safe characters and add padding
  const base64 = str
    .replace(/-/g, '+')
    .replace(/_/g, '/')
    .padEnd(str.length + ((4 - (str.length % 4)) % 4), '=');
  return Buffer.from(base64, 'base64');
}
